// @wordpress/components/build/toolbar/toolbar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Toolbar = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classnames = _interopRequireDefault(require("classnames"));

var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));

var _toolbarGroup = _interopRequireDefault(require("../toolbar-group"));

var _toolbarContainer = _interopRequireDefault(require("./toolbar-container"));

var _context = require("../../ui/context");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
// TODO:
// - (optional) make the legacy `DropdownMenu` read the context variable
// - swap the legacy `DropdownMenu` with the new version of the component
//   once it's stable
const CONTEXT_SYSTEM_VALUE = {
  DropdownMenu: {
    // Note: the legacy `DropdownMenu` component is not yet reactive to this
    // context variant. See https://github.com/WordPress/gutenberg/pull/51097.
    variant: 'toolbar'
  }
};

function UnforwardedToolbar({
  className,
  label,
  ...props
}, ref) {
  if (!label) {
    (0, _deprecated.default)('Using Toolbar without label prop', {
      since: '5.6',
      alternative: 'ToolbarGroup component',
      link: 'https://developer.wordpress.org/block-editor/components/toolbar/'
    });
    return (0, _element.createElement)(_toolbarGroup.default, (0, _extends2.default)({}, props, {
      className: className
    }));
  } // `ToolbarGroup` already uses components-toolbar for compatibility reasons.


  const finalClassName = (0, _classnames.default)('components-accessible-toolbar', className);
  return (0, _element.createElement)(_context.ContextSystemProvider, {
    value: CONTEXT_SYSTEM_VALUE
  }, (0, _element.createElement)(_toolbarContainer.default, (0, _extends2.default)({
    className: finalClassName,
    label: label,
    ref: ref
  }, props)));
}
/**
 * Renders a toolbar.
 *
 * To add controls, simply pass `ToolbarButton` components as children.
 *
 * ```jsx
 * import { Toolbar, ToolbarButton } from '@wordpress/components';
 * import { formatBold, formatItalic, link } from '@wordpress/icons';
 *
 * function MyToolbar() {
 *   return (
 *     <Toolbar label="Options">
 *       <ToolbarButton icon={ formatBold } label="Bold" />
 *       <ToolbarButton icon={ formatItalic } label="Italic" />
 *       <ToolbarButton icon={ link } label="Link" />
 *     </Toolbar>
 *   );
 * }
 * ```
 */


const Toolbar = (0, _element.forwardRef)(UnforwardedToolbar);
exports.Toolbar = Toolbar;
var _default = Toolbar;
exports.default = _default;
