// @wordpress/block-editor/build/components/global-styles/effects-panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EffectsPanel;
exports.useHasEffectsPanel = useHasEffectsPanel;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _utils = require("./utils");

var _object = require("../../utils/object");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function useHasEffectsPanel(settings) {
  const hasShadowControl = useHasShadowControl(settings);
  return hasShadowControl;
}

function useHasShadowControl(settings) {
  return !!(settings !== null && settings !== void 0 && settings.shadow);
}

function EffectsToolsPanel(_ref) {
  let {
    resetAllFilter,
    onChange,
    value,
    panelId,
    children
  } = _ref;

  const resetAll = () => {
    const updatedValue = resetAllFilter(value);
    onChange(updatedValue);
  };

  return (0, _element.createElement)(_components.__experimentalToolsPanel, {
    label: (0, _i18n.__)('Effects'),
    resetAll: resetAll,
    panelId: panelId
  }, children);
}

const DEFAULT_CONTROLS = {
  shadow: true
};

function EffectsPanel(_ref2) {
  let {
    as: Wrapper = EffectsToolsPanel,
    value,
    onChange,
    inheritedValue = value,
    settings,
    panelId,
    defaultControls = DEFAULT_CONTROLS
  } = _ref2;

  const decodeValue = rawValue => (0, _utils.getValueFromVariable)({
    settings
  }, '', rawValue); // Shadow


  const hasShadowEnabled = useHasShadowControl(settings);
  const shadow = decodeValue(inheritedValue === null || inheritedValue === void 0 ? void 0 : inheritedValue.shadow);

  const setShadow = newValue => {
    onChange((0, _object.immutableSet)(value, ['shadow'], newValue));
  };

  const hasShadow = () => !!(value !== null && value !== void 0 && value.shadow);

  const resetShadow = () => setShadow(undefined);

  const resetAllFilter = (0, _element.useCallback)(previousValue => {
    return { ...previousValue,
      shadow: undefined
    };
  }, []);
  return (0, _element.createElement)(Wrapper, {
    resetAllFilter: resetAllFilter,
    value: value,
    onChange: onChange,
    panelId: panelId
  }, hasShadowEnabled && (0, _element.createElement)(_components.__experimentalToolsPanelItem, {
    label: (0, _i18n.__)('Shadow'),
    hasValue: hasShadow,
    onDeselect: resetShadow,
    isShownByDefault: defaultControls.shadow,
    panelId: panelId
  }, (0, _element.createElement)(_components.__experimentalItemGroup, {
    isBordered: true,
    isSeparated: true
  }, (0, _element.createElement)(ShadowPopover, {
    shadow: shadow,
    onShadowChange: setShadow,
    settings: settings
  }))));
}

const ShadowPopover = _ref3 => {
  let {
    shadow,
    onShadowChange,
    settings
  } = _ref3;
  const popoverProps = {
    placement: 'left-start',
    offset: 36,
    shift: true
  };
  return (0, _element.createElement)(_components.Dropdown, {
    popoverProps: popoverProps,
    className: "block-editor-global-styles-effects-panel__shadow-dropdown",
    renderToggle: renderShadowToggle(),
    renderContent: () => (0, _element.createElement)(_components.__experimentalDropdownContentWrapper, {
      paddingSize: "medium"
    }, (0, _element.createElement)(ShadowPopoverContainer, {
      shadow: shadow,
      onShadowChange: onShadowChange,
      settings: settings
    }))
  });
};

function renderShadowToggle() {
  return _ref4 => {
    let {
      onToggle,
      isOpen
    } = _ref4;
    const toggleProps = {
      onClick: onToggle,
      className: (0, _classnames.default)({
        'is-open': isOpen
      }),
      'aria-expanded': isOpen
    };
    return (0, _element.createElement)(_components.Button, toggleProps, (0, _element.createElement)(_components.__experimentalHStack, {
      justify: "flex-start"
    }, (0, _element.createElement)(_icons.Icon, {
      className: "block-editor-global-styles-effects-panel__toggle-icon",
      icon: _icons.shadow,
      size: 24
    }), (0, _element.createElement)(_components.FlexItem, null, (0, _i18n.__)('Shadow'))));
  };
}

function ShadowPopoverContainer(_ref5) {
  var _settings$shadow, _settings$shadow$pres, _settings$shadow2, _settings$shadow2$pre, _settings$shadow3;

  let {
    shadow,
    onShadowChange,
    settings
  } = _ref5;
  const defaultShadows = settings === null || settings === void 0 ? void 0 : (_settings$shadow = settings.shadow) === null || _settings$shadow === void 0 ? void 0 : (_settings$shadow$pres = _settings$shadow.presets) === null || _settings$shadow$pres === void 0 ? void 0 : _settings$shadow$pres.default;
  const themeShadows = settings === null || settings === void 0 ? void 0 : (_settings$shadow2 = settings.shadow) === null || _settings$shadow2 === void 0 ? void 0 : (_settings$shadow2$pre = _settings$shadow2.presets) === null || _settings$shadow2$pre === void 0 ? void 0 : _settings$shadow2$pre.theme;
  const defaultPresetsEnabled = settings === null || settings === void 0 ? void 0 : (_settings$shadow3 = settings.shadow) === null || _settings$shadow3 === void 0 ? void 0 : _settings$shadow3.defaultPresets;
  const shadows = [...(defaultPresetsEnabled ? defaultShadows : []), ...(themeShadows || [])];
  return (0, _element.createElement)("div", {
    className: "block-editor-global-styles-effects-panel__shadow-popover-container"
  }, (0, _element.createElement)(_components.__experimentalVStack, {
    spacing: 4
  }, (0, _element.createElement)(_components.__experimentalHeading, {
    level: 5
  }, (0, _i18n.__)('Shadow')), (0, _element.createElement)(ShadowPresets, {
    presets: shadows,
    activeShadow: shadow,
    onSelect: onShadowChange
  })));
}

function ShadowPresets(_ref6) {
  let {
    presets,
    activeShadow,
    onSelect
  } = _ref6;
  return !presets ? null : (0, _element.createElement)(_components.__experimentalGrid, {
    columns: 6,
    gap: 0,
    align: "center",
    justify: "center"
  }, presets.map(_ref7 => {
    let {
      name,
      slug,
      shadow
    } = _ref7;
    return (0, _element.createElement)(ShadowIndicator, {
      key: slug,
      label: name,
      isActive: shadow === activeShadow,
      onSelect: () => onSelect(shadow === activeShadow ? undefined : shadow),
      shadow: shadow
    });
  }));
}

function ShadowIndicator(_ref8) {
  let {
    label,
    isActive,
    onSelect,
    shadow
  } = _ref8;
  return (0, _element.createElement)("div", {
    className: "block-editor-global-styles-effects-panel__shadow-indicator-wrapper"
  }, (0, _element.createElement)(_components.Button, {
    className: "block-editor-global-styles-effects-panel__shadow-indicator",
    onClick: onSelect,
    label: label,
    style: {
      boxShadow: shadow
    },
    showTooltip: true
  }, isActive && (0, _element.createElement)(_icons.Icon, {
    icon: _icons.check
  })));
}
