// @wordpress/block-editor/build-module/components/global-styles/effects-panel.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { __experimentalToolsPanel as ToolsPanel, __experimentalToolsPanelItem as ToolsPanelItem, __experimentalItemGroup as ItemGroup, __experimentalHStack as HStack, __experimentalVStack as VStack, __experimentalGrid as Grid, __experimentalHeading as Heading, FlexItem, Dropdown, __experimentalDropdownContentWrapper as DropdownContentWrapper, Button } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { useCallback } from '@wordpress/element';
import { shadow as shadowIcon, Icon, check } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import { getValueFromVariable } from './utils';
import { immutableSet } from '../../utils/object';
export function useHasEffectsPanel(settings) {
  const hasShadowControl = useHasShadowControl(settings);
  return hasShadowControl;
}

function useHasShadowControl(settings) {
  return !!(settings !== null && settings !== void 0 && settings.shadow);
}

function EffectsToolsPanel(_ref) {
  let {
    resetAllFilter,
    onChange,
    value,
    panelId,
    children
  } = _ref;

  const resetAll = () => {
    const updatedValue = resetAllFilter(value);
    onChange(updatedValue);
  };

  return createElement(ToolsPanel, {
    label: __('Effects'),
    resetAll: resetAll,
    panelId: panelId
  }, children);
}

const DEFAULT_CONTROLS = {
  shadow: true
};
export default function EffectsPanel(_ref2) {
  let {
    as: Wrapper = EffectsToolsPanel,
    value,
    onChange,
    inheritedValue = value,
    settings,
    panelId,
    defaultControls = DEFAULT_CONTROLS
  } = _ref2;

  const decodeValue = rawValue => getValueFromVariable({
    settings
  }, '', rawValue); // Shadow


  const hasShadowEnabled = useHasShadowControl(settings);
  const shadow = decodeValue(inheritedValue === null || inheritedValue === void 0 ? void 0 : inheritedValue.shadow);

  const setShadow = newValue => {
    onChange(immutableSet(value, ['shadow'], newValue));
  };

  const hasShadow = () => !!(value !== null && value !== void 0 && value.shadow);

  const resetShadow = () => setShadow(undefined);

  const resetAllFilter = useCallback(previousValue => {
    return { ...previousValue,
      shadow: undefined
    };
  }, []);
  return createElement(Wrapper, {
    resetAllFilter: resetAllFilter,
    value: value,
    onChange: onChange,
    panelId: panelId
  }, hasShadowEnabled && createElement(ToolsPanelItem, {
    label: __('Shadow'),
    hasValue: hasShadow,
    onDeselect: resetShadow,
    isShownByDefault: defaultControls.shadow,
    panelId: panelId
  }, createElement(ItemGroup, {
    isBordered: true,
    isSeparated: true
  }, createElement(ShadowPopover, {
    shadow: shadow,
    onShadowChange: setShadow,
    settings: settings
  }))));
}

const ShadowPopover = _ref3 => {
  let {
    shadow,
    onShadowChange,
    settings
  } = _ref3;
  const popoverProps = {
    placement: 'left-start',
    offset: 36,
    shift: true
  };
  return createElement(Dropdown, {
    popoverProps: popoverProps,
    className: "block-editor-global-styles-effects-panel__shadow-dropdown",
    renderToggle: renderShadowToggle(),
    renderContent: () => createElement(DropdownContentWrapper, {
      paddingSize: "medium"
    }, createElement(ShadowPopoverContainer, {
      shadow: shadow,
      onShadowChange: onShadowChange,
      settings: settings
    }))
  });
};

function renderShadowToggle() {
  return _ref4 => {
    let {
      onToggle,
      isOpen
    } = _ref4;
    const toggleProps = {
      onClick: onToggle,
      className: classnames({
        'is-open': isOpen
      }),
      'aria-expanded': isOpen
    };
    return createElement(Button, toggleProps, createElement(HStack, {
      justify: "flex-start"
    }, createElement(Icon, {
      className: "block-editor-global-styles-effects-panel__toggle-icon",
      icon: shadowIcon,
      size: 24
    }), createElement(FlexItem, null, __('Shadow'))));
  };
}

function ShadowPopoverContainer(_ref5) {
  var _settings$shadow, _settings$shadow$pres, _settings$shadow2, _settings$shadow2$pre, _settings$shadow3;

  let {
    shadow,
    onShadowChange,
    settings
  } = _ref5;
  const defaultShadows = settings === null || settings === void 0 ? void 0 : (_settings$shadow = settings.shadow) === null || _settings$shadow === void 0 ? void 0 : (_settings$shadow$pres = _settings$shadow.presets) === null || _settings$shadow$pres === void 0 ? void 0 : _settings$shadow$pres.default;
  const themeShadows = settings === null || settings === void 0 ? void 0 : (_settings$shadow2 = settings.shadow) === null || _settings$shadow2 === void 0 ? void 0 : (_settings$shadow2$pre = _settings$shadow2.presets) === null || _settings$shadow2$pre === void 0 ? void 0 : _settings$shadow2$pre.theme;
  const defaultPresetsEnabled = settings === null || settings === void 0 ? void 0 : (_settings$shadow3 = settings.shadow) === null || _settings$shadow3 === void 0 ? void 0 : _settings$shadow3.defaultPresets;
  const shadows = [...(defaultPresetsEnabled ? defaultShadows : []), ...(themeShadows || [])];
  return createElement("div", {
    className: "block-editor-global-styles-effects-panel__shadow-popover-container"
  }, createElement(VStack, {
    spacing: 4
  }, createElement(Heading, {
    level: 5
  }, __('Shadow')), createElement(ShadowPresets, {
    presets: shadows,
    activeShadow: shadow,
    onSelect: onShadowChange
  })));
}

function ShadowPresets(_ref6) {
  let {
    presets,
    activeShadow,
    onSelect
  } = _ref6;
  return !presets ? null : createElement(Grid, {
    columns: 6,
    gap: 0,
    align: "center",
    justify: "center"
  }, presets.map(_ref7 => {
    let {
      name,
      slug,
      shadow
    } = _ref7;
    return createElement(ShadowIndicator, {
      key: slug,
      label: name,
      isActive: shadow === activeShadow,
      onSelect: () => onSelect(shadow === activeShadow ? undefined : shadow),
      shadow: shadow
    });
  }));
}

function ShadowIndicator(_ref8) {
  let {
    label,
    isActive,
    onSelect,
    shadow
  } = _ref8;
  return createElement("div", {
    className: "block-editor-global-styles-effects-panel__shadow-indicator-wrapper"
  }, createElement(Button, {
    className: "block-editor-global-styles-effects-panel__shadow-indicator",
    onClick: onSelect,
    label: label,
    style: {
      boxShadow: shadow
    },
    showTooltip: true
  }, isActive && createElement(Icon, {
    icon: check
  })));
}
